File: tests/src/main/java/com/github/klee0kai/test/mowgli/animal/Mowgli.java
package com.github.klee0kai.test.mowgli.animal;

import com.github.klee0kai.stone.types.wrappers.Ref;
import com.github.klee0kai.stone.types.wrappers.LazyProvide;
import com.github.klee0kai.stone.types.wrappers.PhantomProvide;
import com.github.klee0kai.test.mowgli.body.Blood;
import com.github.klee0kai.test.mowgli.community.History;
import com.github.klee0kai.test.mowgli.galaxy.Earth;
import com.github.klee0kai.test.mowgli.identity.Conscience;
import com.github.klee0kai.test.mowgli.identity.Knowledge;

import javax.inject.Inject;
import javax.inject.Provider;
import java.lang.ref.SoftReference;
import java.lang.ref.WeakReference;

public class Mowgli implements IAnimal {

    @Inject
    public Blood blood;
    @Inject
    public Earth earth;
    @Inject
    public History history;
    @Inject
    public Conscience conscience;
    @Inject
    public Knowledge knowledge;

    @Inject
    public WeakReference<Knowledge> knowledgeWeakRef;

    @Inject
    public SoftReference<Knowledge> knowledgeSoftRef;

    @Inject
    public LazyProvide<Knowledge> knowledgeLazyProvide;

    @Inject
    public Ref<Knowledge> knowledgePhantomProvide2;

    @Inject
    public Provider<Knowledge> knowledgePhantomProvide3;

    @Inject
    public PhantomProvide<Knowledge> knowledgePhantomProvide;


    public WeakReference<Knowledge> methodKnowledgeWeakRef;

    public SoftReference<Knowledge> methodKnowledgeSoftRef;

    public LazyProvide<Knowledge> methodKnowledgeLazyProvide;

    public Ref<Knowledge> methodKnowledgePhantomProvide2;

    public Provider<Knowledge> methodKnowledgePhantomProvide3;

    public PhantomProvide<Knowledge> methodKnowledgePhantomProvide;

    @Inject
    public void refInject(WeakReference<Knowledge> knowledgeWeakRef, SoftReference<Knowledge> knowledgeSoftRef) {
        methodKnowledgeWeakRef = knowledgeWeakRef;
        methodKnowledgeSoftRef = knowledgeSoftRef;
    }


    @Inject
    public void wrapperInject(
            LazyProvide<Knowledge> knowledgeLazyProvide,
            Ref<Knowledge> knowledgePhantomProvide2,
            Provider<Knowledge> knowledgePhantomProvide3,
            PhantomProvide<Knowledge> knowledgePhantomProvide
    ) {
        methodKnowledgeLazyProvide = knowledgeLazyProvide;
        methodKnowledgePhantomProvide2 = knowledgePhantomProvide2;
        methodKnowledgePhantomProvide3 = knowledgePhantomProvide3;
        methodKnowledgePhantomProvide = knowledgePhantomProvide;
    }


}

File: tests/src/main/java/com/github/klee0kai/test/mowgli/animal/Mowgli.kt
package com.github.klee0kai.test.mowgli.animal

import com.github.klee0kai.stone.weakref.Ref
import com.github.klee0kai.stone.wrappers.LazyProvide
import com.github.klee0kai.stone.wrappers.PhantomProvide
import com.github.klee0kai.test.mowgli.body.Blood
import com.github.klee0kai.test.mowgli.community.History
import com.github.klee0kai.test.mowgli.galaxy.Earth
import com.github.klee0kai.test.mowgli.identity.Conscience
import com.github.klee0kai.test.mowgli.identity.Knowledge
import java.lang.ref.SoftReference
import java.lang.ref.WeakReference
import javax.inject.Inject
import javax.inject.Provider

class Mowgli : IAnimal {

    @Inject
    var blood: Blood? = null

    @Inject
    var earth: Earth? = null

    @Inject
    var history: History? = null

    @Inject
    var conscience: Conscience? = null

    @JvmField
    @Inject
    var knowledge: Knowledge? = null

    @JvmField
    @Inject
    var knowledgeWeakRef: WeakReference<Knowledge?>? = null

    @JvmField
    @Inject
    var knowledgeSoftRef: SoftReference<Knowledge?>? = null

    @JvmField
    @Inject
    var knowledgeLazyProvide: LazyProvide<Knowledge?>? = null

    @JvmField
    @Inject
    var knowledgePhantomProvide2: Ref<Knowledge?>? = null

    @Inject
    var knowledgePhantomProvide3: Provider<Knowledge?>? = null

    @JvmField
    @Inject
    var knowledgePhantomProvide: PhantomProvide<Knowledge?>? = null

    @JvmField
    var methodKnowledgeWeakRef: WeakReference<Knowledge?>? = null

    @JvmField
    var methodKnowledgeSoftRef: SoftReference<Knowledge?>? = null

    @JvmField
    var methodKnowledgeLazyProvide: LazyProvide<Knowledge?>? = null

    @JvmField
    var methodKnowledgePhantomProvide2: Ref<Knowledge?>? = null

    @JvmField
    var methodKnowledgePhantomProvide3: Provider<Knowledge?>? = null

    @JvmField
    var methodKnowledgePhantomProvide: PhantomProvide<Knowledge?>? = null

    @Inject
    fun refInject(knowledgeWeakRef: WeakReference<Knowledge?>?, knowledgeSoftRef: SoftReference<Knowledge?>?) {
        methodKnowledgeWeakRef = knowledgeWeakRef
        methodKnowledgeSoftRef = knowledgeSoftRef
    }


    @Inject
    fun wrapperInject(
        knowledgeLazyProvide: LazyProvide<Knowledge?>?,
        knowledgePhantomProvide2: Ref<Knowledge?>?,
        knowledgePhantomProvide3: Provider<Knowledge?>?,
        knowledgePhantomProvide: PhantomProvide<Knowledge?>?
    ) {
        methodKnowledgeLazyProvide = knowledgeLazyProvide
        methodKnowledgePhantomProvide2 = knowledgePhantomProvide2
        methodKnowledgePhantomProvide3 = knowledgePhantomProvide3
        methodKnowledgePhantomProvide = knowledgePhantomProvide
    }
}
